src/psyclone/psyir/nodes/array_mixin.py

```diff
@@ -243,6 +244,80 @@ def get_lbound_expression(self, pos):
         return BinaryOperation.create(BinaryOperation.Operator.LBOUND, ref,
                                       Literal(str(pos+1), INTEGER_TYPE))
 
+    def get_ubound_expression(self, pos):
+        '''
+        Lookup the upper bound of this ArrayMixin. If we don't have the
+        necessary type information then a call to the UBOUND intrinsic is
+        constructed and returned.
+
+        :param int pos: the dimension of the array for which to lookup the \
+                        upper bound.
+
+        :returns: the declared upper bound for the specified dimension of \
+            the array accesed or a call to the UBOUND intrinsic if it is \
+            unknown.
+        :rtype: :py:class:`psyclone.psyir.nodes.Node`
+
+        '''
+        # First, walk up to the parent reference and get its type. For a simple
+        # ArrayReference this will just be self.
+        root_ref = self.ancestor(Reference, include_self=True)
+        cursor_type = root_ref.symbol.datatype
+
+        # Walk back down the structure, looking up the type information as we
+        # go. We also collect the necessary information for creating a new
+        # Reference as argument to the UBOUND intrinsic in case the type
+        # information is not available.
+        cnames = []
+        cursor = root_ref
+        while cursor is not self:
+            cursor = cursor.member
+            # Collect member information.
+            if isinstance(cursor, ArrayMixin):
+                new_indices = [idx.copy() for idx in cursor.indices]
+                cnames.append((cursor.name.lower(), new_indices))
+            else:
+                cnames.append(cursor.name.lower())
+            # Continue to resolve datatype unless we hit an
+            # UnknownType or DeferredType.
+            if isinstance(cursor_type, ArrayType):
+                cursor_type = cursor_type.intrinsic
+            if isinstance(cursor_type, DataTypeSymbol):
+                cursor_type = cursor_type.datatype
+            if isinstance(cursor_type, (UnknownType, DeferredType)):
+                continue
+            cursor_type = cursor_type.components[cursor.name.lower()].datatype
+
+        if (isinstance(cursor_type, ArrayType) and
+                cursor_type.shape[pos] not in [ArrayType.Extent.DEFERRED,
+                                               ArrayType.Extent.ATTRIBUTE]):
+            # We have the full type information and the upper bound is known.
+            return cursor_type.shape[pos].upper.copy()
```

Comment:
Is this new method identical to lbound but for this line and L318? Given the complexity of walking the tree to discover type information, it would be best to avoid duplicating the code. We could keep the current interfaces (`get_u/lbound_expression`) but have them as simple wrappers to a call to an internal method that takes an extra argument specifying which bound is required.

Reply:
Done